processing/src/main/java/org/apache/druid/query/filter/ArrayContainsElementFilter.java

```diff
@@ -0,0 +1,538 @@
+      return input -> {
+        if (input == null) {
+          return false;
+        }
+        boolean anyMatch = false;
+        for (Object elem : input) {
+          anyMatch = anyMatch || elementComparator.compare(elem, matchVal) == 0;
+        }
+        return anyMatch;
+      };
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      ArrayContainsPredicateFactory that = (ArrayContainsPredicateFactory) o;
+      if (!Objects.equals(elementMatchValue.type(), that.elementMatchValue.type())) {
+        return false;
+      }
+      if (elementMatchValue.isArray()) {
+        return Arrays.deepEquals(elementMatchValue.asArray(), that.elementMatchValue.asArray());
+      }
+      return Objects.equals(elementMatchValue.value(), that.elementMatchValue.value());
+    }
+
+    @Override
+    public int hashCode()
+    {
+      return Objects.hash(elementMatchValue);
+    }
+  }
+
+  private static class TypedConstantElementValueMatcherFactory extends EqualityFilter.TypedConstantValueMatcherFactory
+  {
+    public TypedConstantElementValueMatcherFactory(
+        ExprEval<?> matchValue,
+        DruidPredicateFactory predicateFactory
+    )
+    {
+      super(matchValue, predicateFactory);
+    }
+
+    @Override
+    public ValueMatcher makeDimensionProcessor(DimensionSelector selector, boolean multiValue)
+    {
+      if (matchValue.valueOrDefault() == null || matchValue.isArray()) {
+        return predicateMatcherFactory.makeDimensionProcessor(selector, multiValue);
+      }
+      return super.makeDimensionProcessor(selector, multiValue);
+    }
+
+    @Override
+    public ValueMatcher makeFloatProcessor(BaseFloatColumnValueSelector selector)
+    {
+      if (matchValue.valueOrDefault() == null || matchValue.isArray()) {
+        return predicateMatcherFactory.makeFloatProcessor(selector);
+      }
+      return super.makeFloatProcessor(selector);
+    }
+
+    @Override
+    public ValueMatcher makeDoubleProcessor(BaseDoubleColumnValueSelector selector)
+    {
+      if (matchValue.valueOrDefault() == null || matchValue.isArray()) {
+        return predicateMatcherFactory.makeDoubleProcessor(selector);
+      }
+      return super.makeDoubleProcessor(selector);
+    }
+
+    @Override
+    public ValueMatcher makeLongProcessor(BaseLongColumnValueSelector selector)
+    {
+      if (matchValue.valueOrDefault() == null || matchValue.isArray()) {
+        return predicateMatcherFactory.makeLongProcessor(selector);
+      }
+      return super.makeLongProcessor(selector);
+    }
+
+    @Override
+    public ValueMatcher makeArrayProcessor(
+        BaseObjectColumnValueSelector<?> selector,
+        @Nullable ColumnCapabilities columnCapabilities
+    )
+    {
+      return super.makeArrayProcessor(selector, columnCapabilities);
```

Comment:
while it kinda does the right thing right now, it feels too tightly coupled with EqualityFilter. It confused me for some time till I realized that you are passing the predicateMatcherFactory to super class. The way I read it as that elementMatchValue can be compared to input array through "equals" comparison. I think it would be better to just call  `predicateMatcherFactory.makeArrayProcessor(selector, columnCapabilities);` and in makeComplexProcessor

Reply:
> while it kinda does the right thing right now, it feels too tightly coupled with EqualityFilter
that is the point though, if the filter is matching a scalar column and the match value is a non-null scalar value then this filter should behave identical to equality filter, the ask during review of #15350 was to re-use as much of the equality filter as possible because in certain cases it is the equality filter

Reply:
actually, i just removed these overrides since they were pointless